pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/BrokerRegistryImpl.java

```diff
@@ -0,0 +1,302 @@
+        this.brokerLookupDataLockManager = pulsar.getCoordinationService().getLockManager(BrokerLookupData.class);
+        this.scheduler = pulsar.getLoadManagerExecutor();
+        this.brokerLookupDataCache = new ConcurrentHashMap<>();
+        this.listeners = new ArrayList<>();
+        this.cacheReloadFuture = CompletableFuture.completedFuture(null);
+        this.lookupServiceAddress = pulsar.getLookupServiceAddress();
+        this.brokerZNodePath = LOOKUP_DATA_PATH + "/" + lookupServiceAddress;
+        this.brokerLookupData = new BrokerLookupData(
+                pulsar.getSafeWebServiceAddress(),
+                pulsar.getWebServiceAddressTls(),
+                pulsar.getBrokerServiceUrl(),
+                pulsar.getBrokerServiceUrlTls(),
+                pulsar.getAdvertisedListeners(),
+                pulsar.getProtocolDataToAdvertise(),
+                pulsar.getConfiguration().isEnablePersistentTopics(),
+                pulsar.getConfiguration().isEnableNonPersistentTopics(),
+                pulsar.getBrokerVersion());
+    }
+
+    @Override
+    public void start() throws PulsarServerException {
+        if (started.get()) {
+            return;
+        }
+        pulsar.getLocalMetadataStore().registerListener(this::handleMetadataStoreNotification);
+        try {
+            this.register();
+            // Update all lookup data to cache
+            this.reloadAllBrokerLookupCacheAsync()
+                    .get(conf.getMetadataStoreOperationTimeoutSeconds(), TimeUnit.SECONDS);
+            this.started.set(true);
+        } catch (MetadataStoreException | ExecutionException | InterruptedException | TimeoutException e) {
+            throw new PulsarServerException(e);
+        }
+    }
+
+    @Override
+    public boolean isStarted() {
+        return this.started.get();
+    }
+
+    @Override
+    public synchronized void register() throws MetadataStoreException {
+        if (!registered) {
+            try {
+                this.brokerLookupDataLock = brokerLookupDataLockManager.acquireLock(brokerZNodePath, brokerLookupData)
+                        .get(conf.getMetadataStoreOperationTimeoutSeconds(), TimeUnit.SECONDS);
+                registered = true;
+            } catch (InterruptedException | ExecutionException | TimeoutException e) {
+                throw MetadataStoreException.unwrap(e);
+            }
+        }
+    }
+
+    @Override
+    public synchronized void unregister() throws MetadataStoreException {
+        if (registered) {
+            try {
+                brokerLookupDataLock.release().get(conf.getMetadataStoreOperationTimeoutSeconds(), TimeUnit.SECONDS);
+                registered = false;
+            } catch (CompletionException | InterruptedException | ExecutionException | TimeoutException e) {
+                throw MetadataStoreException.unwrap(e);
+            }
+        }
+    }
+
+    @Override
+    public String getLookupServiceAddress() {
```

Comment:
I have been thinking about changing the method name to `getBrokerId()`, since the function of `LookupServiceAddress` is the unique identifier.

@heesung-sn What do you think?

Reply:
There is no description about the broker ID in the proposal. It might make readers confused. For example, an ID is usually an integer, e.g. the broker ID in Kafka.

Could `getBrokerId()` return a unique string that does not represent the address? If not, I think keeping the `getLookupServiceAddress` not changed would be better,

Reply:
Use `brokerId` as the name may be more generic? Since the `getBrokerId ` method depends on the implementation. And it can use `LookupServiceAddress` as the broker id, or we can support other formats of broker id in the future if necessary.